packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -338,6 +347,10 @@ export function renderWithHooks(
   refOrContext: any,
   nextRenderExpirationTime: ExpirationTime,
 ): any {
+  if (__DEV__) {
+    Component = resolveTypeForHotReloading(Component);
```

Comment:
This makes conceptual sense but it means we have to remember not to read `fiber.type` anywhere else. We faced a similar problem with `React.lazy`, which is one of the reasons we split the `type` field into `elementType` (the thing we compare during reconciliation) and `type` (the implementation of the render phase).

Let's move this to `cloneFiber` and `createFiber` and assign the resolved type to the `fiber.type` field.

Reply:
How confident are we that those will always run before we do some work? Are there any cases at all where they wouldn't? Stale types make me nervous.

My justification for this was that I only ever want to support hot reloading for one thing — rendering function components. And rendering ultimately happens here.

For example I don't want to accidentally mess with types for classes. Or even for resolving default props. There's also a bunch of `fiber.type === fiber.elementType` checks that add extra work like resolving, and I hoped not to cause those codepaths to fork after a reload. Seems more confusing.

How confident are you that this would be better?

Reply:
Fixed.